Modify rule S946: Expand and adjust for LaYC (#3163)
SonarSource/rspec: rules/S946/cfamily/rule.adoc

> Assigning the address of an automatic object to another object that outlives it is dangerous since the address is invalidated after the first object has ceased to exist.

Reviewer: An explanation of what an "automatic object" is could be useful here.

Author: That's a good point. I'll try to craft a brief description.

> Assigning the address of an automatic object to another object that outlives it is dangerous since the address is invalidated after the first object has ceased to exist.

Reviewer: ```suggestion
Persisting the address of an automatic object after it ceases to exist is dangerous because any attempt to dereference it would result in undefined behavior.
```
I think what makes this description harder to read is the fact that many sentences have two different "object" in them that need to be fully qualified each time to make sense.

Reviewer 2: +1, but I would recommend avoiding the term "dereference" as we can also have dangling references.
Have you considered the term "access"? See https://godbolt.org/z/43c3esze1

== Why is this an issue?

An automatic object is an object whose lifetime is automatically managed.

> The storage for an automatic object, e.g. a local variable, is allocated at the beginning of the enclosing code block and is deallocated at the end.

Reviewer 2: ```suggestion
The storage for an automatic object, e.g. a local variable, is allocated at the beginning of the enclosing code block and is deallocated at the end.
This is commonly referred to as "allocated on the stack".
```
I think it's useful to make a connection to the common name of this so that the user could connect this to other resources.

> If the address of an automatic object is assigned to another automatic object of larger scope, a static or extern object, or if it is returned from a function (using `return` or an output parameter), then the object containing the address may exist beyond the time at which the original object ceases to exist.

Reviewer: Just a suggestion:
```suggestion
If the address of an automatic object is assigned to another automatic object of larger scope, a static or extern object, or if it is returned from a function (using `return` or an output parameter), then there will be a point where the address will point to an object that ceased to exist.
```

> In that case, the address becomes invalid and attempts to dereference the invalid address result in undefined behavior.

Reviewer 2: ```suggestion
In that case, the address becomes invalid, and attempts to access it is undefined behavior.
```

[source,cpp]
----
int *global = nullptr;

int* bar(int **out) {
  int local = 42;
  int *ptr;
  global = &local; // Noncompliant: assigning the address of an object allocated on the stack to a global variable
  {
    int i = 9001;
    ptr = &i; // Noncompliant: assigning the address of a stack-allocated object to an object that outlives it
  }
  *out = &local; // Noncompliant: returning the address of an object allocated on the stack (via output parameter)
  return &local; // Noncompliant: returning the address of an object allocated on the stack

> }

Reviewer 2: It's unfortunate, but we don't detect escapes through out parameters.
And I was surprised that we don't raise S6655 for `ptr`. Then I realized that we would need to read or write (in short: access) the dangling pointer `ptr` to make use raise the issue.

I'd recommend raising a ticket for the out parameter and reworking the example.

Author: Interesting. I'll create a ticket for that and introduce lit tests in the corresponding `sonar-cpp` pull request that links to the ticket -- that's how Tomek and I handled such situations previously.

Author: cf. https://sonarsource.atlassian.net/browse/CPP-4734

----

== What is the potential impact?

> Persisting addresses of objects with automatic storage using objects that outlive them is dangerous as it may create invalid addresses.

Reviewer: ```suggestion
Persisting addresses of objects with automatic storage past their lifetime is dangerous as it may create invalid addresses.
```

Attempts to access such invalidated addresses result in undefined behavior.

For programs that exercise undefined behavior, the compiler is no longer bound by the language specification.
The application may crash or, even worse, the application may appear to execute correctly while losing data or producing incorrect results.

== How to fix it

> Do not persist addresses of objects with automatic storage using objects that outlive them.

Reviewer: This sentence only repeats what the title says. Summing up the below solutions might be more useful:

```suggestion
There are multiple approaches to avoid persisting addresses of object with automatic storage after their lifetime ended:

* Make a full copy of the object instead of using an address.
* Allocate the address on the heap and deallocate it once done.
* Create the automatic object in a larger scope where it outlives all storages of its address.
* Reduce the scope of objects that store the address of the original automatic object.
```

}
----

==== Compliant solution

[source,cpp,diff-id=1,diff-type=compliant]
----
int foo() {
  int local_auto = 42;
  return local_auto; // Compliant: returns a value rather than an address
}
----

[source,cpp,diff-id=1,diff-type=compliant]
----
int* foo() {
  int *local_auto = new int(42);
  return local_auto; // Compliant: returns the address of a heap-allocated object
}
----

==== Noncompliant code example

[source,cpp,diff-id=2,diff-type=noncompliant]
----
const char *str_ptr = nullptr;

void bar() {
  const char str[] = "This will change";
  str_ptr = str; // Noncompliant: address of `str` becomes invalid when `str` goes out of scope
}
----

==== Compliant solution

[source,cpp,diff-id=2,diff-type=compliant]
----
void bar() {
  const char str[] = "This will change";
  const char *str_ptr = str; // Compliant: `str_ptr` has the same storage duration as `str`
}
----

==== Noncompliant code example

[source,cpp,diff-id=3,diff-type=noncompliant]
----
int *buz() {
  int array[256];
  for (size_t i = 0; i < std::size(array); ++i) {
    array[i] = i;
  }
  return array; // Noncompliant: returns address of stack-allocated object

> }

Reviewer 2: It's unfortunate, but due to loop unrolling of known bounds limitations, see CPP-3607, we won't analyze anything after the loop statement, thus we never reach the return where we could raise this issue.

Author: At least we already have a ticket for that. I'll rework that code a bit, since we are not really interested in analyzing loops here.

==== Compliant solution

[source,cpp,diff-id=2,diff-type=compliant]
----
void bar() {
  const char str[] = "This will change";
  const char *str_ptr = str; // Compliant: `str_ptr` has the same storage duration as `str`
}
----

==== Noncompliant code example

[source,cpp,diff-id=3,diff-type=noncompliant]
----
int *buz() {
  int array[256];
  for (size_t i = 0; i < 256; ++i) {
    array[i] = i;
  }
  return array; // Noncompliant: returns address of stack-allocated object
}
----

==== Compliant solution

[source,cpp,diff-id=3,diff-type=compliant]
----
void buz(int *array, size_t length) {
  for (size_t i = 0; i < length; ++i) {
    array[i] = i;
  }
}

void caller() {
  int array[256];
  buz(array, sizeof(array) / sizeof(array[0]));
}
----

==== Noncompliant code example

[source,cpp,diff-id=4,diff-type=noncompliant]
----
void fun(char **out) {
  char buffer[64];
  for (size_t i = 0; i < length; ++i) {
    buffer[i] = 42;
  }
  *out = buffer; // Noncompliant: `buffer`'s address becomes invalid once it goes out of scope
}

void caller() {
  char *p;
  fun(&p);
}
----

==== Compliant solution

[source,cpp,diff-id=4,diff-type=compliant]
----
char buffer[64];

void fun(char **out) {

>   for (size_t i = 0; i < sizeof(buffer) / sizeof(buffer[0]); ++i) {

Reviewer: ```suggestion
  for (char &c: buffer) {
```
or
```suggestion
  std::fill(std::begin(buffer), std::end(buffer), 42);
```

  global = &local; // Noncompliant: assigning the address of an object allocated on the stack to a global variable
  {
    int i = 9001;
    ptr = &i; // Noncompliant: assigning the address of a stack-allocated object to an object that outlives it
  }
  *out = &local; // Noncompliant: returning the address of an object allocated on the stack (via output parameter)
  return &local; // Noncompliant: returning the address of an object allocated on the stack
}
----

== What is the potential impact?

Persisting addresses of objects with automatic storage past their lifetime is dangerous as it creates invalid addresses.
Attempts to access objects that no longer exist through such invalid addresses result in undefined behavior.

For programs that exercise undefined behavior, the compiler is no longer bound by the language specification.
The application may crash or, even worse, the application may appear to execute correctly while losing data or producing incorrect results.

== How to fix it

There are multiple approaches to avoid persisting addresses of objects with automatic storage after their lifetime ended:

* Make a full copy of the object instead of using its address.

> * Allocate the object on the heap and deallocate it once done.

Reviewer 2: Depending on the book, it's sometimes called "heap", and othertimes called "free-store".
I believe the Scott Meyers and Bjarne books differ on this taxonomy. Maybe it's better to xref these.
```suggestion
* Allocate the object on the heap (free-store) and deallocate it once done.
```

I found a paper actually using the "free-store" terminology: https://wg21.link/P1950.

  *out = buffer; // Noncompliant: `buffer`'s address becomes invalid once it goes out of scope
}

void caller() {
  char *p;
  fun(&p);
}
----

==== Compliant solution

[source,cpp,diff-id=4,diff-type=compliant]
----
#include <algorithm>
#include <array>

char buffer[64];

void fun(char **out) {
  std::fill(std::begin(buffer), std::end(buffer), 42);
  *out = buffer; // Compliant: `buffer`'s lifetime is the program's lifetime
}

void caller() {
  char *p;
  fun(&p);
}
----

[source,cpp,diff-id=4,diff-type=compliant]
----
#include <vector>

std::vector<int> fun() {
  std::vector<int> buf(/* count */ 64, /* initial value */ 42);
  return buf;
}

void caller() {
  auto buf = fun();
}
----

== Resources

=== Conference presentations

* CppCon 2018 - https://www.youtube.com/watch?v=uQyT-5iWUow&ab_channel=CppCon[Surprises in Object Lifetime]

=== Standards

* CERT - https://wiki.sei.cmu.edu/confluence/x/UtcxBQ[DCL30-C. Declare objects with appropriate storage durations]
* CERT - https://wiki.sei.cmu.edu/confluence/display/cplusplus/EXP54-CPP.+Do+not+access+an+object+outside+of+its+lifetime[EXP54-CPP. Do not access an object outside of its lifetime]

> * CERT - https://wiki.sei.cmu.edu/confluence/display/c/MSC00-C.+Compile+cleanly+at+high+warning+levels[MSC00-C. Compile cleanly at high warning levels]

Reviewer 2: For these CERT links, sometimes its's a short link, sometimes a long link.
Please consolidate this and use the format that is used across different rules.

Author: Thanks for pointing this out. By majority vote (using `grep`) we now use short links 😁
